rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,489 @@
+    cb_tx: mpsc::Sender<IpcServerMsg>,
+}
+
+impl Callbacks for CallbackHandlerIpc {
+    fn on_disconnect(&self, _error: &connlib_client_shared::Error) {
+        self.cb_tx
+            .try_send(IpcServerMsg::OnDisconnect)
+            .expect("should be able to send OnDisconnect");
+    }
+
+    fn on_set_interface_config(
+        &self,
+        ipv4: Ipv4Addr,
+        ipv6: Ipv6Addr,
+        dns: Vec<IpAddr>,
+    ) -> Option<i32> {
+        tracing::info!("TunnelReady");
+        self.cb_tx
+            .try_send(IpcServerMsg::OnSetInterfaceConfig { ipv4, ipv6, dns })
+            .expect("Should be able to send TunnelReady");
+        None
+    }
+
+    fn on_update_resources(&self, resources: Vec<ResourceDescription>) {
+        tracing::info!(len = resources.len(), "New resource list");
+        self.cb_tx
+            .try_send(IpcServerMsg::OnUpdateResources(resources))
+            .expect("Should be able to send OnUpdateResources");
+    }
+}
+
+async fn handle_ipc_client(server: named_pipe::NamedPipeServer) -> Result<()> {
+    let framed = Framed::new(server, LengthDelimitedCodec::new());
+    let mut framed = pin!(framed);
+    let (cb_tx, mut cb_rx) = mpsc::channel(100);
+
+    let mut connlib = None;
+    let callback_handler = CallbackHandlerIpc { cb_tx };
+    loop {
+        let ev = poll_fn(|cx| {
+            match cb_rx.poll_recv(cx) {
+                Poll::Ready(Some(msg)) => return Poll::Ready(Ok(IpcEvent::Connlib(msg))),
+                Poll::Ready(None) => {
+                    return Poll::Ready(Err(anyhow!(
+                        "Impossible - MPSC channel from connlib closed"
+                    )))
+                }
+                Poll::Pending => {}
+            }
+
+            match framed.as_mut().poll_next(cx) {
+                Poll::Ready(Some(msg)) => {
+                    let msg = serde_json::from_slice(&msg?)?;
+                    return Poll::Ready(Ok(IpcEvent::Client(msg)));
+                }
+                Poll::Ready(None) => return Poll::Ready(Ok(IpcEvent::IpcDisconnect)),
+                Poll::Pending => {}
+            }
+
+            Poll::Pending
+        })
+        .await;
+
+        match ev {
+            Ok(IpcEvent::Client(msg)) => match msg {
+                IpcClientMsg::Connect { api_url, token } => {
+                    let token = secrecy::SecretString::from(token);
+                    assert!(connlib.is_none());
+                    let device_id = connlib_shared::device_id::get()
+                        .context("Failed to read / create device ID")?;
+                    let (private_key, public_key) = keypair();
+
+                    let login = LoginUrl::client(
+                        Url::parse(&api_url)?,
+                        &token,
+                        device_id.id,
+                        None,
+                        public_key.to_bytes(),
+                    )?;
+
+                    // TODO: Configurable max partition time?
```

Comment:
Just need to make sure connlib won't remove the token if the laptop goes offline for a sustained period of time (e.g. airplane)

Reply:
So it will still be considered signed in because there's a token, even if you're offline for hours?

Reply:
Yeah. On the Apple client we actually sleep it, but connlib just keeps connected in the background for a long time